_posts/2025-05-05-Exposing-WASM-binaries-as-MCP-tools.adoc

```diff
@@ -0,0 +1,125 @@
+
+WebAssembly (WASM) has emerged as a powerful technology for running high-performance code in various environments. In this article, we'll explore how to expose WASM binaries as Model Context Protocol (MCP) tools in WildFly.
+With WildFly AI Feature Pack 0.5.0 we have introduced a new subsystem to expose WASM WASI binaries via CDI, and since we also support MCP tools declaration via CDI, just exposing those binaries as MCP tools was a small step forward.
+
+== What are MCP tools?
+
+The Model Context Protocol (MCP) defines a standard way for systems (like an IDE or a CLI) to interact with MCP servers. These servers expose *tools* (functions that can be executed remotely with specific inputs) and *resources* (data sources that can be accessed). This allows the client system to leverage the server's capabilities and data for various tasks.
+
+== Benefits of WASM-based MCP tools
+
+* Near-native performance for compute-intensive operations
+* Ability to reuse existing tools written in languages like C, C++, or Rust
+* Secure execution within WildFly's sandbox environment
+* Cross-platform compatibility without recompilation
+
+== Defining WASM Binaries for MCP Tool Integration
+
+Before a WASM binary can be exposed as an MCP tool, it needs to be defined within WildFly's WASM subsystem. This subsystem acts as a registry for your WASM modules.
+It is based on the https://github.com/extism/chicory-sdk[chicory-sdk], an experimental Java SDK using the https://github.com/dylibso/chicory[JVM-native Chicory] runtime.
+
+=== Configuration Methods
+
+You can configure these binaries using either the WildFly CLI or by directly editing the XML configuration (`standalone.xml`).
+
+==== Using CLI
+
+Add a WASM binary definition via the CLI like this:
+
+[source,bash]
+----
+/subsystem=wasm/wasm-tool=my-wasm-tool:add(path="/path/to/your/binary.wasm")
+----
+
+*   `my-wasm-tool`: A unique name for this binary definition.
+*   `path`: The location of the `.wasm` file. This path is relative to the `relative-to` attribute if specified.
+
+==== Using XML Configuration
+
+Alternatively, add the definition within the `<subsystem xmlns="urn:wildfly:wasm:1.0">` element in your configuration file:
+
+[source,xml]
+----
+<subsystem xmlns="urn:jboss:domain:wasm:1.0">
+    <wasm-tool name="my-wasm-tool" path="path/relative/to/module/root/binary.wasm"/>
+    <!-- Other binary definitions -->
+</subsystem>
+----
+
+Ensure the `path` is correctly specified relative to the root of the deployment module.
+
+== Next Steps: CDI Integration
+
+Once defined in the WASM subsystem, these binaries can be injected using CDI.
+The first way you can use them is to get them by name as an `org.wildfly.wasm.api.WasmInvoker` using the `@WasmTool` annotation.
+This very basic interface wraps te call to the binary, passing the parameters as a byte array and getting the result as a byte array.
+
+[source,java]
+----
+
+    @Inject
+    @WasmTool(value = "greet")
```

Comment:
is that related to the configuration example above? Would I pass the `my-wasm-tool` to use my configured WASM binary?

Reply:
Yes, I've updated the example code to reflect that